### ISSUE TITLE
Disappearing nodes

### ISSUE THREAD

**Issue by 314eter** [tweag/topiary#489]
### Problem

Topiary assumes that all content is contained in leaf nodes. Without that assumption it's difficult to recreate the original content from the syntax tree, but tree-sitter was never created with that use case in mind, so I expect there to be a lot of tree-sitter grammars that break this assumption.

It's mostly fine in tree-sitter-ocaml, but there are some cases where adding a comment in a leaf node can make the comment the leaf node instead.

### Example

```ocaml
type '(**)a x = [`(**)X] [@x(**).x]
```
will be formatted as
```ocaml
type (**) x = [ (**) ] [@ (**) ]
```

Most grammars have content outside of leaf nodes in strings too if there are escape sequences, but that's not a problem as long as they are captured with `@leaf`.

### Possible solutions

1. Enforce the "content should be in leaf nodes" rule on all tree-sitter grammars used by topiary. It can be hard to find and fix the corner cases, but for tree-sitter-ocaml it's feasible.
2. Add special handling for comments (and other extras) in topiary. This will probably cover most of the problematic cases. Relates to #430.

**Comment by ErinvanderVeen** [comment]
Hej! Thanks for this issue! We have someone currently investigating #430, so I hope issues like this will be resolved after that work is done.

We have contributed some upstream patches in cases where we could not work around issues we encountered, but I don't think we can ask grammar maintainers to cater to Topiary specifically (yet). Also, like you mentioned, this might be doable for OCaml, but there are probably many other languages were this is well out of scope.

In short, I hope resolving #430 will resolve issues like this, and otherwise we will try to think of some other solution that does not involve many upstream patches! If that too fails, we can still start with upstream patches.

**Comment by nbacquey** [comment]
After a bit of investigation, it appears that a third option could be to tag as `@leaf` all nodes that are leaves in practice unless a comment is inserted inside.

I'm not a fan of this one, and I think the work I'm doing to close #430 will also solve this issue.

**Comment by 314eter** [comment]
Tagging them as `@leaf` will indeed prevent topiary from outputting invalid code. Downside is that it will not reformat something like
```ocaml
type ' a x = [` X] [@x . x]
```
to
```ocaml
type 'a x = [`X] [@x.x]
```
But even if #430 is solved, that will be difficult without having separate nodes for `'` and `a`.
